### PR TITLE
{AzureContainerApp} az containerapp create --bind "my-app" throws error when service has dashes in the name "-" 

### PR DESCRIPTION
**Pull request by navba-MSFT**
fixes Azure/azure-cli-extensions#6524

Container Apps lets me create a container with dashes ("-") in the name, but then I can't reference that container in the --bind parameter. so I expect the command to accept the service with a dash in the name because I was allowed to create it in the first place.

```
cli.azure.cli.core.azclierror: Traceback (most recent call last):
  File "D:\a\_work\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\knack/cli.py", line 233, in invoke
  File "D:\a\_work\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/core/commands/__init__.py", line 663, in execute
  File "D:\a\_work\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/core/commands/__init__.py", line 726, in _run_jobs_serially
  File "D:\a\_work\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/core/commands/__init__.py", line 718, in _run_job
  File "C:\Users\charris\.azure\cliextensions\containerapp\azext_containerapp\_client_factory.py", line 28, in _polish_bad_errors
    raise ex
  File "D:\a\_work\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/core/commands/__init__.py", line 697, in _run_job
  File "D:\a\_work\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/core/commands/__init__.py", line 333, in __call__
  File "D:\a\_work\1\s\build_scripts\windows\artifacts\cli\Lib\site-packages\azure/cli/core/commands/command_operation.py", line 121, in handler
  File "C:\Users\charris\.azure\cliextensions\containerapp\azext_containerapp\custom.py", line 627, in create_containerapp
    service_connectors_def_list, service_bindings_def_list = parse_service_bindings(cmd, service_bindings,
  File "C:\Users\charris\.azure\cliextensions\containerapp\azext_containerapp\_utils.py", line 542, in parse_service_bindings
    raise InvalidArgumentValueError("The Binding Name can only contain letters, numbers (0-9), periods ('.'), "
azure.cli.core.azclierror.InvalidArgumentValueError: The Binding Name can only contain letters, numbers (0-9), periods ('.'), and underscores ('_'). The length must not be more than 60 characters.

cli.azure.cli.core.azclierror: The Binding Name can only contain letters, numbers (0-9), periods ('.'), and underscores ('_'). The length must not be more than 60 characters.
az_command_data_logger: The Binding Name can only contain letters, numbers (0-9), periods ('.'), and underscores ('_'). The length must not be more than 60 characters.
cli.knack.cli: Event: Cli.PostExecute [<function AzCliLogging.deinit_cmd_metadata_logging at 0x048C0580>]

```

---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [ ] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally?

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`.

### CODE REVIEW

**Comment by yonzhan** [comment]
AzureContainerApp

**Comment by zhoxing-ms** [comment]
@Greedygre Could you please help review this PR?

File: src/containerapp/HISTORY.rst [hunk 1]
```diff
@@ -2,6 +2,10 @@
 
 Release History
 ===============
+0.3.37
```
Comment (Greedygre) [inline src/containerapp/HISTORY.rst:5]: I think you should use `upcoming`, and not release a new version 0.3.37 in this PR.

File: src/containerapp/setup.py [hunk 2]
```diff
@@ -28,7 +28,7 @@
 # TODO: Confirm this is the right version number you want and it matches your
 # HISTORY.rst entry.
 
-VERSION = '0.3.36'
+VERSION = '0.3.37'
```
Comment (Greedygre) [inline src/containerapp/setup.py:31]: Please not release a new version, we'll release periodically.

**Comment by Greedygre** [comment]
Hi @yash-nisar 
Could you please help to review this PR?

File: src/containerapp/azext_containerapp/_utils.py [hunk 3]
```diff
@@ -467,7 +467,7 @@ def process_service(cmd, resource_list, service_name, arg_dict, subscription_id,
 
 
 def validate_binding_name(binding_name):
-    pattern = r'^(?=.{1,60}$)[a-zA-Z0-9._]+$'
+    pattern = r'^(?=.{1,60}$)[a-zA-Z0-9._-]+$'
```
[the remark this answers is not in the document]
Comment (Greedygre) [reply src/containerapp/azext_containerapp/_utils.py:470]: Hi @yash-nisar Do you know whether the binding name can contains `-` ?

File: src/containerapp/HISTORY.rst (hunk 1 above)
Comment (Greedygre) [inline src/containerapp/HISTORY.rst:5]:
```suggestion
upcoming
```

Comment (navba-MSFT) [reply src/containerapp/HISTORY.rst]: done

**Comment by navba-MSFT** [comment]
@yash-nisar @Greedygre Please let me know if you had a chance to look into this.

**Comment by Greedygre** [comment]
Hi @navba-MSFT 
Please add test for the --bind with dash in the name.

File: src/containerapp/azext_containerapp/_utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Greedygre) [reply src/containerapp/azext_containerapp/_utils.py:470]: Hi @navba-MSFT 
I have test --bind with dash in name. It works fine.
You need to update the `raise InvalidArgumentValueError` .

**Comment by yash-nisar** [comment]
Currently, while using --bind, we expect the user to provide us the svc_name:[binding_name] where the binding_name is optional and the binding_name should follow 

"The Binding Name can only contain letters, numbers (0-9), periods ('.'), and underscores ('_'). The length must not be more than 60 characters." because this is what service linker expects.

If we have a svc_name that has a "-", we use the same as our binding_name and ask the user to provide an acceptable binding_name. Since binding_names would appear in env vars, we cannot allow "-" in the binding_name plus service linker doesn't accept it.

CC @duglin

**Comment by yash-nisar** [comment]
There were talks going on about how to handle it and there is an issue on the same: https://github.com/orgs/serverless-paas-balam/projects/2/views/1?pane=issue&itemId=34057347

**Comment by charris-msft** [comment]
From a DevEx perspective, customers already face significant challenges naming azure resources. Some allow non-alphanumeric chars and some don't some allow uppercase, and some don't. It's the wild west out there.

The current model, allowing them to give something a name that will prevent future functionality is obviously broken. Ideally, I would like to see "Azure" support the same naming conventions on every service, so customers don't have to have special knowledge of a resource in order to provide a name for it. But I fear that would lead us to only allowing lowercase, alphanumeric chars which are harder for humans to read.

I'm also concerned about the idea of "restricting dev service names", but perhaps it'd due to a lack of understanding on my part. I need to better understand what exactly a "dev service" refers to because it seems to me what I'm passing to --bind is just the name of a container app.

I do like the idea of supporting a compliant binding name that differs from the container app name, and I believe most developers wouldn't have a problem with swapping the invalid "-" char for "_". But I would also give them the opportunity to specify the binding name themself - perhaps with an optional `--bindng_name` param that allows them to be explicit if needed.

**Comment by duglin** [comment]
> But I would also give them the opportunity to specify the binding name themself - perhaps with an optional --bindng_name param that allows them to be explicit if needed.

That’s already supported today on the --bind flag: --bind SERVICE_NAME[:BINDING_NAME]
It’s just optional.

From: Chris Harris ***@***.***>
Date: Tuesday, July 25, 2023 at 11:35 AM
To: Azure/azure-cli-extensions ***@***.***>
Cc: Doug Davis ***@***.***>, Mention ***@***.***>
Subject: Re: [Azure/azure-cli-extensions] {AzureContainerApp} az containerapp create --bind "my-app" throws error when service has dashes in the name "-" (PR #6530)

From a DevEx perspective, customers already face significant challenges naming azure resources. Some allow non-alphanumeric chars and some don't some allow uppercase, and some don't. It's the wild west out there.

The current model, allowing them to give something a name that will prevent future functionality is obviously broken. Ideally, I would like to see "Azure" support the same naming conventions on every service, so customers don't have to have special knowledge of a resource in order to provide a name for it. But I fear that would lead us to only allowing lowercase, alphanumeric chars which are harder for humans to read.

I'm also concerned about the idea of "restricting dev service names", but perhaps it'd due to a lack of understanding on my part. I need to better understand what exactly a "dev service" refers to because it seems to me what I'm passing to --bind is just the name of a container app.

I do like the idea of supporting a compliant binding name that differs from the container app name, and I believe most developers wouldn't have a problem with swapping the invalid "-" char for "_". But I would also give them the opportunity to specify the binding name themself - perhaps with an optional --bindng_name param that allows them to be explicit if needed.

**Comment by charris-msft** [comment]
ok, I feel a bit foolish for not figuring that out, but in my feeble defense it doesn't just leap off the page.

Perhaps a simple fix, while the team considers other options, is to just update the error message to let the user know they can specify their own binding name:

from this:
```
The Binding Name can only contain letters, numbers (0-9), periods ('.'), and underscores ('_'). The length must not be more than 60 characters.
``` 
to this:
```
The Binding Name can only contain letters, numbers (0-9), periods ('.'), and underscores ('_'). The length must not be more than 60 characters. 
By default, the binding name is the same as the service name you specified [my-aca-pgaddon], but you can override the default and specify your own compliant binding name like this --bind my-aca-pgaddon[:my_aca_pgaddon]
```

image

**Comment by navba-MSFT** [comment]
> 

I will close this PR for now and created the below PR to address this ask. 
https://github.com/Azure/azure-cli-extensions/pull/6566

**Pull request closed without merging**